isos/appliance-xtra.sh

```diff
@@ -0,0 +1,185 @@
+#!/bin/bash
+# Copyright 2016 VMware, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Build the appliance filesystem ontop of the base
+set -x
+
+# exit on failure and configure debug, include util functions
+set -e && [ -n "$DEBUG" ] && set -x
+DIR=$(dirname $(readlink -f "$0"))
+. $DIR/base/utils.sh
+
+
+function usage() {
+echo "Usage: $0 -p staged-package(tgz) -b binary-dir -x binary-source -f binary-filename (inside the ISO) -o appliance-output-name" 1>&2
+exit 1
+}
+
+while getopts "p:b:x:f:o:" flag
+do
+    case $flag in
+
+        p)
+            # Required. Package name
+            PACKAGE="$OPTARG"
+            ;;
+
+        b)
+            # Required. Target for iso and source for components
+            BIN="$OPTARG"
+            ;;
+
+        x)
+			# Required. Source of the xtra binary to add to the ISO
+            XTRABIN="$OPTARG"
+            ;;
+
+        f)
+			# Required. Filename of the xtra binary inside the ISO
+            XTRABIN_FILENAME="$OPTARG"
+            ;;
+
+        o)
+			APPLIANCE_OUTNAME="$OPTARG"
+			;;
+
+        *)
+            usage
+            ;;
+    esac
+done
+
+shift $((OPTIND-1))
+
+# check there were no extra args and the required ones are set
+if [ ! -z "$*" -o -z "$PACKAGE" -o -z "${BIN}" ]; then
+    usage
+fi
+
+echo XTRABIN: ${XTRABIN}
+echo XTRABIN_FILENAME: ${XTRABIN_FILENAME}
+
+if [ -z "${XTRABIN}" ] || [ -z "${XTRABIN_FILENAME}" ]; then
+	usage
+fi
+
+if [ -n "${XTRABIN}" ] && [ -z "${XTRABIN_FILENAME}" ]; then
+	echo "-f binary-filename  must be specificed when selecting -x"
+	usage
+fi
+
+PKGDIR=$(mktemp -d)
+
+# unpackage base package
+unpack $PACKAGE $PKGDIR
+
+#################################################################
+# Above: arg parsing and setup
+# Below: the image authoring
+#################################################################
+
+echo RootFsDir $(rootfs_dir $PKGDIR)
+
+# sysctl
+cp ${DIR}/appliance/sysctl.conf $(rootfs_dir $PKGDIR)/etc/
```

Comment:
Starting from here, this duplicates substantial non-trivial image authoring code from `appliance.sh`. We should have a single copy of this code so that we don't wind up with inconsistencies if they drift out of sync.